crates/joystick/src/linux.rs

```diff
@@ -180,10 +220,95 @@ impl Joystick {
             self.instant = Instant::now();
         }
     }
+
+    fn press(&mut self, button: KeyCode) {
+        let time = self.instant.elapsed();
+
+        self.events.push(JoystickEvent {
+            key: button,
+            event_type: EventType::KEY,
+            duration: time,
+            action: KeyAction::Press,
+        });
+    }
+
+    fn release(&mut self, button: KeyCode) {
+        let time = self.instant.elapsed();
+
+        self.events.push(JoystickEvent {
+            key: button,
+            event_type: EventType::KEY,
+            duration: time,
+            action: KeyAction::Release,
+        });
+    }
+
+    fn release_later(&mut self, button: KeyCode, duration: Duration) {
+        let time = self.instant.elapsed();
+
+        self.events.push(JoystickEvent {
+            key: button,
+            event_type: EventType::KEY,
+            duration: time + duration,
+            action: KeyAction::Release,
+        });
+    }
+
+    fn tap(&mut self, button: KeyCode) {
+        // send in press and release
+        let release_duration = Duration::from_millis(TAP_DURATION);
+        self.press(button);
+        self.release_later(button, release_duration);
+    }
+}
+
+impl Default for Joystick {
+    fn default() -> Self {
+        let name = "Future TAS Joystick Linux";
+        let abs_setup = AbsInfo::new(256, 0, 512, 20, 20, 1);
+        let abs_x = UinputAbsSetup::new(AbsoluteAxisCode::ABS_X, abs_setup);
+        let abs_y = UinputAbsSetup::new(AbsoluteAxisCode::ABS_Y, abs_setup);
+        let mut keys = AttributeSet::<KeyCode>::new();
+
+        keys.insert(KeyCode::BTN_DPAD_DOWN);
+        keys.insert(KeyCode::BTN_DPAD_LEFT);
+        keys.insert(KeyCode::BTN_DPAD_RIGHT);
+        keys.insert(KeyCode::BTN_DPAD_UP);
+        keys.insert(KeyCode::BTN_NORTH);
+        keys.insert(KeyCode::BTN_SOUTH);
+        keys.insert(KeyCode::BTN_EAST);
+        keys.insert(KeyCode::BTN_WEST);
+        keys.insert(KeyCode::BTN_START);
+        keys.insert(KeyCode::BTN_SELECT);
+        keys.insert(KeyCode::BTN_TR);
+        keys.insert(KeyCode::BTN_TL);
+        keys.insert(KeyCode::BTN_TR2);
+        keys.insert(KeyCode::BTN_TL2);
+
+        let device = VirtualDeviceBuilder::new()
```

Comment:
This one is a mouthful O_o

Reply:
unfortunately its the api for the evdev wrapper XD